hook-common/src/pgqueue.rs

```diff
@@ -0,0 +1,215 @@
+use std::str::FromStr;
+
+use chrono::prelude::*;
+use serde::{de::DeserializeOwned, Serialize};
+use sqlx::postgres::{PgPool, PgPoolOptions};
+use thiserror::Error;
+
+/// Enumeration of errors for operations with PgQueue.
+/// Errors can originate from sqlx and are wrapped by us to provide additional context.
+#[derive(Error, Debug)]
+pub enum PgQueueError {
+    #[error("connection failed with: {error}")]
+    ConnectionError {
+        error: sqlx::Error
+    },
+    #[error("{command} query failed with: {error}")]
+    QueryError {
+        command: String,
+        error: sqlx::Error
+    },
+    #[error("{0} is not a valid JobStatus")]
+    ParseJobStatusError(String),
+}
+
+/// Enumeration of possible statuses for a Job.
+/// Available: A job that is waiting in the queue to be picked up by a worker.
+/// Completed: A job that was successfully completed by a worker.
+/// Failed: A job that was unsuccessfully completed by a worker.
+/// Running: A job that was picked up by a worker and it's currentlly being run.
+#[derive(Debug, PartialEq, sqlx::Type)]
+#[sqlx(type_name = "job_status")]
+#[sqlx(rename_all = "lowercase")]
+pub enum JobStatus {
+    Available,
+    Completed,
+    Failed,
+    Running,
+}
+
+/// Allow casting JobStatus from strings.
+impl FromStr for JobStatus {
+    type Err = PgQueueError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "available" => Ok(JobStatus::Available),
+            "completed" => Ok(JobStatus::Completed),
+            "failed" => Ok(JobStatus::Failed),
+            "running" => Ok(JobStatus::Running),
+            invalid => Err(PgQueueError::ParseJobStatusError(invalid.to_owned())),
+        }
+    }
+}
+
+/// JobParameters are stored and read to and from a JSONB field, so we accept anything that fits `sqlx::types::Json`.
+pub type JobParameters<J> = sqlx::types::Json<J>;
+
+/// A Job to be executed by a worker dequeueing a PgQueue.
+#[derive(sqlx::FromRow)]
+pub struct Job<J> {
+    pub id: i64,
+    pub attempt: i32,
+    pub finished_at: Option<DateTime<Utc>>,
+    pub created_at: DateTime<Utc>,
+    pub started_at: Option<DateTime<Utc>>,
```

Comment:
We can always add it later, so feel free to ignore me, but I think the [attempted_by](https://github.com/riverqueue/river/blob/a8f0ce74e93ee3c73f91b2781cbe0033af2a8554/internal/dbsqlc/river_job.sql#L16) field in River would be super useful for debugging (i.e. it could be the pod name, or something)

Reply:
Yeah, works for me, could even be thread id in a single pod context.